### ISSUE TITLE
几个在 SwiftUI 中使用惰性容器的技巧和注意事项

### ISSUE THREAD

**Issue by fatbobman** [fatbobman/blogComments#225]
几个在 SwiftUI 中使用惰性容器的技巧和注意事项

在 SwiftUI 的框架中，惰性布局容器，如 List 和 LazyVStack，提供了一种高效展示大型数据集的方法。这些容器的设计精妙，它们仅在必要时才动态地构建和加载视图，从而显著优化了应用的性能和内存使用效率。本文将探讨一些实用技巧和重要注意事项，旨在赋予开发者利用 SwiftUI 惰性容器时增强应用响应性和资源管理的能力。

Tips and Considerations for Using Lazy Containers in SwiftUI

In the SwiftUI framework, lazy layout containers such as List and LazyVStack provide a method for efficiently displaying large datasets. These containers are ingeniously designed to dynamically build and load views only when necessary, thereby significantly optimizing the app's performance and memory efficiency. This article will explore some practical tips and important considerations aimed at enhancing developers' ability to leverage SwiftUI's lazy containers for improved app responsiveness and resource management.

**Comment by jihongboo** [comment]
使用RandomAccessCollection来触发加载更多确实是一个新的思路，虽然比直接在最后一个cell上加block要麻烦一点，但是处理的空间更灵活，可以实现还未滚动到最后一个cell之前就提前加载的逻辑。稍微封装一下就能很好用。

**Comment by yanbodang** [comment]
uiImage = UIImage(data: data) 会引起主线程的卡顿。

**Comment by astralchen** [comment]
在iOS 18.1上运行，items 发生变化，下面代码不会重新渲染：

 List {
      ForEach(dataSource) { item in
        Text("\(item.number)")
      }
    }

需要改成：

List(dataSource) { item in
    Text("\(item.number)")
}

**Comment by fatbobman** [comment]
@motafan 典型的 Bug，给苹果提 FB 吧

**Comment by chadbag** [comment]
The problem with the infinite data loading shown is that it does not allow true random access.  If an index is put into subscript that is outside the loaded data then you either return a nil or you have to stall while you load more data.  This would be something like the anti-pattern of using a semaphore and then a Task to load and wait on the Task to finish before returning out of subscript with the data.

Also, you constantly append new data but never purge old data that gets further away from the current window of data.  So eventually you've loaded the whole data set into memory.

Ideally your backing data would support only having a window into data that can move across the data, purging from both ends when it's no longer relevant.  But this messes up scrolling when you start to pull the low end data (data before the current viewed window) out of the list of data that you looped over.  The scroll position gets messed up.   Both with a ScrollView/LazyVStack and a List.

Any suggestions on hoot have a true infinite scrolling capability that goes both forward and backwards and only stores the data in the vicinity your displaying, purging from both ends as necessary.  (Maybe you have 1000 elements, display 15 at a time, and keep 45 in memory at a time).  Without messing up the scrolling and scroll position?

I thought have having my "data" just be an array of index values (1...1000 in the above example).  The data store is @Observable with the actual Dictionary backing it up [Int: ValueStruct] being a property of it.  This works and I get good scrolling, except when I scroll too fast and my index ends up outside of the data.  I correctly fetch new data, and temporarily pass back a stand-in object to display while teh data is loading.  But I have not gotten the already displayed views to update when the data load finished for that range.  The Datastore object is @Observable and is @State in the app and is passed to the view which encapsulates the "Scrollview/LazyVStack" as Environment.  And is passed again as Environment to the DataView that is shown inside the ForEach of the LazyVStack.  This DataView has an @State for the ValueStruct to be shown in the view, and that ValueStruct is looked up in .onAppear of the view elements that use it inside DataView.  I'd hoped that all that state "chaining" would cause it to update but it doesn't.

Any suggestions on how to infinite scroll both directions only caching a small widow of data to support the currently visible cells and the immediate page or two above and below?

**Comment by fatbobman** [comment]
@chadbag Implementing truly smooth bidirectional infinite scrolling with dynamic data loading in SwiftUI presents a fundamental challenge rooted in its core design philosophy: state-driven UI. In this architecture, state changes immediately trigger UI reconstruction, preventing the precise timing control available in imperative frameworks.
Specifically, when we add data to any position in an array (whether at the beginning or end), SwiftUI immediately recalculates the entire scroll container's layout, which inevitably affects the current scroll position. In UIKit, we can calculate the content height difference after data updates but before view reconstruction, then manually adjust the contentOffset to maintain visual continuity—this kind of imperative precision control is not feasible within SwiftUI's declarative paradigm.
The new APIs introduced in iOS 17, such as onScrollPhaseChange and onScrollGeometryChange, do provide enhanced scroll control capabilities. Combined with scrollPosition, they can mitigate this issue to some extent. However, due to fundamental architectural differences, the smoothness still falls short of what UIKit can achieve.

**Comment by chadbag** [comment]
@fatbobman  Yes, thank you.  Adding on to the end, the scrollPosition works pretty well to keep you on track.  Removing data from the front is what doesn't work so well, and I had hopes there was a trick I had missed.  Thanks for your reply.   My trick of making the data just an array of the indices of the actual data (which is really an OrderedDictionary so that we can have an index range that moves) works and scrolls smoothly and quickly both directions, though it's tricky updating from async data loads when you scroll past your window.   I have it working, mostly.   Again, thanks for your blog and your comments to my question.
